pkg/segment/writer/segstore.go

```diff
@@ -1037,50 +1036,72 @@ func (ss *SegStore) flushBloomIndex(cname string, bi *BloomIndex) uint64 {
 		return 0
 	}
 
-	defer bffd.Close()
-
-	var buf bytes.Buffer
-	bufWriter := bufio.NewWriter(&buf)
-
 	// there is no accurate way to find exactly how many bytes the write.to is going to write
 	// and we need that number , so that we write it first and then the actual bloom data
 	// hence this messiness to write it to some buffer, get the bytesWritten count and then do
 	// the actual write
-	bytesWritten, bferr := bi.Bf.WriteTo(bufWriter)
-	if bferr != nil {
-		log.Errorf("flushBloomIndex: write buf failed fname=%v, err=%v", fname, bferr)
-		return 0
-	}
-
-	bytesWritten += utils.LEN_BLKNUM_CMI_SIZE // for blkNum
-	bytesWritten += 1                         // reserver for CMI_Type
+	bytesWritten := uint32(0)
 
-	// copy the size of blockBloom in uint32
-	if _, err = bffd.Write(toputils.Uint32ToBytesLittleEndian(uint32(bytesWritten))); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+	_, err = bffd.Write([]byte{0, 0, 0, 0})
+	if err != nil {
+		log.Errorf("flushBloomIndex: failed to skip bytes for bloom size fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += 4
 
 	// copy the blockNum
 	if _, err = bffd.Write(toputils.Uint16ToBytesLittleEndian(ss.numBlocks)); err != nil {
-		log.Errorf("flushBloomIndex: bloomsize write failed fname=%v, err=%v", fname, err)
+		log.Errorf("flushBloomIndex: block num write failed fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += utils.LEN_BLKNUM_CMI_SIZE
 
 	// write CMI type
 	if _, err = bffd.Write(utils.CMI_BLOOM_INDEX); err != nil {
 		log.Errorf("flushBloomIndex: CMI Type write failed fname=%v, err=%v", fname, err)
 		return 0
 	}
+	bytesWritten += 1
 
 	// write the blockBloom
-	_, bferr = bi.Bf.WriteTo(bffd)
-	if bferr != nil {
-		log.Errorf("flushBloomIndex: write blockbloom failed fname=%v, err=%v", fname, bferr)
+	bloomSize, err := bi.Bf.WriteTo(bffd)
+	if err != nil {
+		log.Errorf("flushBloomIndex: write blockbloom failed fname=%v, err=%v", fname, err)
+		return 0
+	}
+	bytesWritten += uint32(bloomSize)
+
+	err = bffd.Close()
+	if err != nil {
+		log.Errorf("flushBloomIndex: close failed fname=%v, err=%v", fname, err)
+		return 0
+	}
+
+	bffd, err = os.OpenFile(fname, os.O_WRONLY, 0644)
```

Comment:
Also, I am curious as to why there is a need to `close` and `open` the file again to write the bytes written. Can you not do `seek` or use `WriteAt` function after writing the bloom data to write the bytes written value and then do `Sync()` and `close`?  

Is there a reason why you have done this way?

Reply:
Seek is not defined and writeAt is not supported for files opened with O_APPEND flag. That's why I am first append the data and then reopening the file with Write only mode and writing the number of bytes written at its correct position.